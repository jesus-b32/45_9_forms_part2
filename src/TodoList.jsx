import React, {useState} from "react";
import NewTodoForm from "./NewTodoForm";
import Todo from "./Todo";

function TodoList () {
    // todos is an array of todo objects
    const [todos, setTodos] = useState([]);

    // update state by adding current todos and newTodo, made from form data, to setState
    function addTodo(newTodo) {
        setTodos(todos => [...todos, newTodo]);
    }

    // filter out todo with matching id
    function removeTodo(id) {
        setTodos(todos => todos.filter(todo => todo.id !== id));
    }

    // create box compnents from boxes state
    const todoComponents = todos.map(todo => (
        <Todo
            key={todo.id}
            id={todo.id}
            task={todo.task}
            handleRemove={removeTodo}
        />
        )
    )

    return (
        <div>
            <NewTodoForm addTodo={addTodo}/>
            {todoComponents}
        </div>
    )
}

export default TodoList;
