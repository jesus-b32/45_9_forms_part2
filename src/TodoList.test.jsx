import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";
import { describe, it, expect, vi } from "vitest";



describe('TodoList', () => {
    function addTodo(todoList, task = "code") {
        const taskInput = todoList.getByLabelText("Task:");
        fireEvent.change(taskInput, { target: { value: task }});
        const submitButton = todoList.getByText("Add a TODO!");
        fireEvent.click(submitButton);
    }
    
    
    it("renders without crashing", function() {
        render(<TodoList />);
    });
    
    it("matches snapshot", function() {
        const { asFragment } = render(<TodoList />);
        expect(asFragment()).toMatchSnapshot();
    });
    
    it("can add a todo", function() {
        const list = render(<TodoList />);
        addTodo(list);
        
        // expect form to clear and todo to be on the page
        expect(list.getByLabelText("Task:")).toHaveValue("");
        expect(list.getByText("code")).toBeInTheDocument();
        expect(list.getByText("Remove Todo")).toBeInTheDocument();
    });
    
    it("can delete a todo", function() {
        const list = render(<TodoList />);
        addTodo(list);
        
        fireEvent.click(list.getByText("Remove Todo"));
        
        // expect todo to be gone
        expect(list.queryByText("code")).not.toBeInTheDocument();
    });
})
