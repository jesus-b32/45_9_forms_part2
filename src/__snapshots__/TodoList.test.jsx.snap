// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoList > matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <label
          for="task"
        >
          Task: 
        </label>
        <input
          id="task"
          name="task"
          type="text"
          value=""
        />
        <button
          id="newTodoButton"
        >
          Add a TODO!
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
